benchmark-framework/src/main/java/io/openmessaging/benchmark/Benchmark.java

```diff
@@ -159,52 +160,62 @@ public static void main(String[] args) throws Exception {
+                        worker.stopAll();
 
-                    worker.initializeDriver(new File(driverConfig));
+                        worker.initializeDriver(new File(driverConfig));
 
-                    WorkloadGenerator generator = new WorkloadGenerator(driverConfiguration.name, workload, worker);
+                        WorkloadGenerator generator = new WorkloadGenerator(driverConfiguration.name, workload, worker);
 
-                    TestResult result = generator.run();
-                    result.beginTime = beginTime;
-                    result.endTime = dateFormat.format(new Date());
-                    result.version = arguments.serviceVersion;
+                        TestResult result = generator.run();
+                        result.beginTime = beginTime;
+                        result.endTime = dateFormat.format(new Date());
+                        result.version = arguments.serviceVersion;
 
-                    boolean useOutput = (arguments.output != null) && (arguments.output.length() > 0);
+                        boolean useOutput = (arguments.output != null) && (arguments.output.length() > 0);
 
-                    String fileName = useOutput? arguments.output: String.format("%s-%s-%s.json", workloadName,
-                    driverConfiguration.name, dateFormat.format(new Date()));
+                        String fileName = useOutput? arguments.output: String.format("%s-%s-%s.json", workloadName,
+                        driverConfiguration.name, dateFormat.format(new Date()));
 
-                    log.info("Writing test result into {}", fileName);
-                    writer.writeValue(new File(fileName), result);
+                        log.info("Writing test result into {}", fileName);
+                        writer.writeValue(new File(fileName), result);
 
-                    generator.close();
-                } catch (Exception e) {
-                    log.error("Failed to run the workload '{}' for driver '{}'", workload.name, driverConfig, e);
-                    // sometimes calling worker.stopAll hangs
-                    // using brute force to exit
-                    System.exit(1);
-                } finally {
-                    try {
-                        worker.stopAll();
-                    } catch (IOException e) {
+                        generator.close();
+                    } catch (Exception e) {
+                        log.error("Failed to run the workload '{}' for driver '{}'", workload.name, driverConfig, e);
+                        Throwables.throwIfUnchecked(e);
+                        throw new RuntimeException(e);
+                    } finally {
+                        try {
+                            worker.stopAll();
+                        } catch (IOException e) {
+                        }
                     }
-                }
+                });
             });
-        });
+            success = true;
```

Comment:
With this pattern I don't need the AtomicBoolean anymore.

I let the exception escape from the nested iteration, and catch it below, but we keep the existing behavior of logging the workload that failed above.